### PR TITLE
[IOTDB-3258] implement count nodes

### PR DESCRIPTION
**Pull request by SpriCoder**
See details: https://issues.apache.org/jira/browse/IOTDB-3258.

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LogicalPlanner.java [hunk 1]
```diff
@@ -423,6 +424,15 @@ public PlanNode visitCountLevelTimeSeries(
           .getRoot();
     }
 
+    @Override
+    public PlanNode visitCountNodes(CountNodesStatement countStatement, MPPQueryContext context) {
+      LogicalPlanBuilder planBuilder = new LogicalPlanBuilder(context);
+      return planBuilder
+          .planChildPathsSchemaSource(countStatement.getPartialPath(), countStatement.getLevel())
+          .planCountMerge()
+          .getRoot();
```
Comment (yifuzhou) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LogicalPlanner.java:433]: How about both sg above and sg below need to count together? like there has sg: root.ln, root.abc.a1.a2.a3 and ts:root.ln.a1.a2.a3. We want count nodes root.** level=3? I think planNodeManagementMemoryMerge() should be used.

File: confignode/src/test/java/org/apache/iotdb/confignode/service/thrift/ConfigNodeRPCServiceProcessorTest.java [hunk 2]
```diff
@@ -1000,4 +1003,61 @@ private void cleanUserAndRole() throws TException {
       Assert.assertEquals(TSStatusCode.SUCCESS_STATUS.getStatusCode(), status.getCode());
     }
   }
+
+  @Test
+  public void testDeleteStorageGroup() throws TException {
```
Comment (yifuzhou) [inline confignode/src/test/java/org/apache/iotdb/confignode/service/thrift/ConfigNodeRPCServiceProcessorTest.java:1008]: It seems that previous pr changes has been removed... This function is the same as deleteStorageGroupTest(). I just change the name. Test class uses ....Test format and test function uses test... format should be more normalized.

**Review by yifuzhou — COMMENTED** [review]
PTAL~

File: confignode/src/test/java/org/apache/iotdb/confignode/service/thrift/ConfigNodeRPCServiceProcessorTest.java (hunk 2 above)
Comment (SpriCoder) [reply confignode/src/test/java/org/apache/iotdb/confignode/service/thrift/ConfigNodeRPCServiceProcessorTest.java:1008]: Ok, get it! I will have a check.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LogicalPlanner.java (hunk 1 above)
Comment (SpriCoder) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LogicalPlanner.java:433]: From my perspective, there is not allowed to have both root.ln and root.ln.a1.a2.a3?

Comment (SpriCoder) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LogicalPlanner.java:433]: Besides, I use get child path in count node, so if there is possible having same path?

**Comment by SpriCoder** [comment]
Already fixed! PTAL @yifuzhou

**Review by yifuzhou — APPROVED** [review]
LGTM!

**Comment by MarcosZyk** [comment]
If the conduction cost of the above suggestion is heave, the suggested simplified modification is that we should not extend the interface of child management with a parameter level, which is confusing, instead we should implement independent node management configRequest and NodeScanOperator.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LogicalPlanBuilder.java [hunk 3]
```diff
@@ -607,20 +607,31 @@ public LogicalPlanBuilder planLevelTimeSeriesCountSource(
     return this;
   }
 
-  public LogicalPlanBuilder planChildPathsSchemaSource(PartialPath partialPath) {
-    this.root = new ChildPathsSchemaScanNode(context.getQueryId().genPlanNodeId(), partialPath);
+  public LogicalPlanBuilder planChildPathsSchemaSource(PartialPath partialPath, Integer level) {
```
Comment (MarcosZyk) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LogicalPlanBuilder.java:610]: Rename this to planNodePathsSchemaSource, or anything without child but related to node management

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LocalExecutionPlanner.java [hunk 4]
```diff
@@ -467,40 +467,51 @@ public Operator visitLevelTimeSeriesCount(
 
     @Override
     public Operator visitChildPathsSchemaScan(
```
Comment (MarcosZyk) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LocalExecutionPlanner.java:469]: Rename to ```visitNodePathsSchemaScan```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/schema/NodePathsConvertOperator.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    for (int i = 0; i < block.getPositionCount(); i++) {
+      String path = block.getColumn(0).getBinary(i).toString();
+      PartialPath partialPath;
+      try {
+        partialPath = new PartialPath(path);
+      } catch (IllegalPathException e) {
+        LOGGER.warn("Failed to convert node path to PartialPath {}", path);
+        continue;
+      }
```
Comment (MarcosZyk) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/schema/NodePathsConvertOperator.java:77]: Hope for an improvement of generic support on TsBlock. Thus the try-catch will be saved.

**Review by MarcosZyk — COMMENTED** [review]
Excellent work! Only a few problems. PTAL.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LocalExecutionPlanner.java (hunk 4 above)
Comment (SpriCoder) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LocalExecutionPlanner.java]: fixed.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LogicalPlanBuilder.java (hunk 3 above)
Comment (SpriCoder) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LogicalPlanBuilder.java]: fixed.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/schema/NodePathsConvertOperator.java (hunk 5 above)
Comment (SpriCoder) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/schema/NodePathsConvertOperator.java:77]: I have the same idea.

**Review by MarcosZyk — APPROVED** [review]

**Pull request merged**